channeldb/graph.go

```diff
@@ -173,16 +174,22 @@ type ChannelGraph struct {
 	cacheMu     sync.RWMutex
 	rejectCache *rejectCache
 	chanCache   *channelCache
+
+	chanScheduler batch.Scheduler
+	nodeScheduler batch.Scheduler
 }
 
 // newChannelGraph allocates a new ChannelGraph backed by a DB instance. The
 // returned instance has its own unique reject cache and channel cache.
 func newChannelGraph(db *DB, rejectCacheSize, chanCacheSize int) *ChannelGraph {
-	return &ChannelGraph{
+	g := &ChannelGraph{
 		db:          db,
 		rejectCache: newRejectCache(rejectCacheSize),
 		chanCache:   newChannelCache(chanCacheSize),
 	}
+	g.chanScheduler = batch.NewTimeScheduler(db.DB, &g.cacheMu, 10*time.Millisecond)
+	g.nodeScheduler = batch.NewTimeScheduler(db.DB, nil, 10*time.Millisecond)
```

Comment:
I was able to set this slightly higher and still achieve the same sync time of about 2 minutes (granted, the peers I synced with on different runs were very fast).  I used as high as 8 seconds, less db txn's.  What do you think?  I think perhaps the two schedulers could be staggered at different intervals so as to not cause db contention?

Reply:
I do see some memory usage with `dereference` calls when growing the mmap since these are large writes if 8 second batch time is used (~10MB) so a case could be made for `InitialMmapSize` as discussed in #3278 , but this is nothing compared to the `GetBlock` calls (6 GB +) made in the router.

Reply:
I also see a reduction from the 10ms scheduler at 845MB:
```
ROUTINE ======================== github.com/lightningnetwork/lnd/batch.(*batch).trigger in /Users/nsa/go/src/github.com/lightningnetwork/lnd/batch/batch.go
         0   845.71MB (flat, cum)  7.09% of Total
         .          .     27:}
         .          .     28:
         .          .     29:// trigger is the entry point for the batch and ensures that run is started at
         .          .     30:// most once.
         .          .     31:func (b *batch) trigger() {
         .   845.71MB     32:	b.start.Do(b.run)
         .          .     33:}
         .          .     34:
         .          .     35:// run executes the current batch of requests. If any individual requests fail
         .          .     36:// alongside others they will be retried by the caller.
         .          .     37:func (b *batch) run() {
```

to 269MB with the 8s scheduler:
```
ROUTINE ======================== github.com/lightningnetwork/lnd/batch.(*batch).trigger in /Users/nsa/go/src/github.com/lightningnetwork/lnd/batch/batch.go
         0   269.59MB (flat, cum)  1.80% of Total
         .          .     27:}
         .          .     28:
         .          .     29:// trigger is the entry point for the batch and ensures that run is started at
         .          .     30:// most once.
         .          .     31:func (b *batch) trigger() {
         .   269.59MB     32:	b.start.Do(b.run)
         .          .     33:}
         .          .     34:
         .          .     35:// run executes the current batch of requests. If any individual requests fail
         .          .     36:// alongside others they will be retried by the caller.
         .          .     37:func (b *batch) run() {
```

Reply:
@Crypt-iQ yes the 10ms time is just a temporary commit so that i could drop in the batched version into the regular API and verify it has the same semantics w/o having to modify any timeouts.

In practice we should be using a timeout on the order of seconds. I did my testing with a duration of 1 second and it performed quite well. We could probably do 8 seconds as well, though this also delays the time till first write pretty significantly. Most of the writes were taking 10-20ms, so only about 2-4% of one core would be dedicated to the db txn itself if we go with 1 second, and even lower with 8.

The main thing left to do is benchmark this on mobile and get a sense for how these parameters behave on different architectures, just haven't gotten around to doing so :)